event/parsing/totalBootTimeParser.go

```diff
@@ -0,0 +1,170 @@
+				return -1, err
+			}
+		}
+	}
+
+	if valid, err := isEventValid(latestPreviousEvent, rebootRegex, time.Now); !valid {
+		level.Error(b.Logger).Log(xlog.ErrorKey(), err)
+		return -1, err
+	}
+
+	// boot time calculation
+	// Event birthdate is saved in unix nanoseconds, so we must first convert it to a unix time using nanoseconds.
+	restartTime := wrpWithTime.Beginning.Sub(time.Unix(0, latestPreviousEvent.BirthDate)).Seconds()
+
+	if restartTime <= 0 {
+		err = errInvalidRestartTime
+		level.Error(b.Logger).Log(xlog.ErrorKey(), err, "Restart time", restartTime)
+		return -1, err
+	}
+
+	return restartTime, nil
+}
+
+// sees if this event is part of the most recent previous session
+func checkLatestPreviousEvent(e Event, previousEventTracked Event, latestBootTime int64, eventType *regexp.Regexp) (Event, error) {
+	eventBootTimeInt, err := GetEventBootTime(e)
+	previousEventBootTime, _ := GetEventBootTime(previousEventTracked)
+
+	if err != nil {
+		return previousEventTracked, err
+	}
+
+	if eventBootTimeInt > latestBootTime {
+		return Event{}, errNewerBootTime
+	}
+
+	// If this event has a boot time greater than what we've seen so far
+	// but still less than the current boot-time, it means that this event
+	// is part of a more recent previous cycle.
+	if eventBootTimeInt > previousEventBootTime && eventBootTimeInt < latestBootTime {
+		return e, nil
+	}
+
+	// If the event is of the type we are looking for and has the same boot-time as the
+	// newest previous boot time, return this event.
+	if eventBootTimeInt == previousEventBootTime && eventType.MatchString(e.Dest) {
+		// If the previously tracked event doesn't match the event we're looking for,
+		// return the current event. If both events match the type of event we are looking for,
+		// compare the birthdates and return the one with the more recent birthdate.
+		if !eventType.MatchString(previousEventTracked.Dest) || e.BirthDate > previousEventTracked.BirthDate {
+			return e, nil
+		}
+	}
+	return previousEventTracked, nil
+}
+
+// Sees if a date is within a certain time frame.
+// PastBuffer should be a positive duration.
+func isDateValid(currTime func() time.Time, pastBuffer time.Duration, futureBuffer time.Duration, date time.Time) bool {
+	if date.Before(time.Unix(0, 0)) || date.Equal(time.Unix(0, 0)) {
+		return false
+	}
+
+	if pastBuffer.Seconds() < 0 {
+		pastBuffer = -1 * pastBuffer
+	}
+
+	now := currTime()
+	pastTime := now.Add(-1 * pastBuffer)
+	futureTime := now.Add(futureBuffer)
+
+	return (pastTime.Before(date) || pastTime.Equal(date)) && (futureTime.Equal(date) || futureTime.After(date))
+}
+
+func isEventValid(event Event, expectedType *regexp.Regexp, currTime func() time.Time) (bool, error) {
+	// see if event found matches expected event type
+	if !expectedType.MatchString(event.Dest) {
+		return false, emperror.WrapWith(errors.New("unexpected event type"), errRestartTime.Error(), "expected type", expectedType.String())
```

Comment:
This `WrapWith()` feels kind of clunky, since we're converting a string to an error and an error to a string....The created error is meant to be `failed to get restart time: unexpected event type`, right?

Maybe you can change `errRestartTime` to a struct that contains an error.  Something like:
```
type errRestartTime error

func (e errRestartTime) Error() string {
  return fmt.Sprintf("failed to get restart time: %v", e)
}
```

Then this code could look something like...
```
return false, emperror.With(errRestartTime(errors.New("unexpected event type")), "expected type", expectedType.String())
```

I haven't thought about this much; there could be a better solution.  Also no guarantees that this code actually works/compiles. 🙂